File: java/client/src/main/java/glide/api/models/commands/json/JsonGetOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands.json;

import lombok.experimental.SuperBuilder;

/**
 * Options for JSON GET command.
 * This class provides configuration options for retrieving JSON values.
 */
@SuperBuilder
public class JsonGetOptions {
    
    /**
     * The path to retrieve from the JSON document.
     * If not specified, the entire document is returned.
     */
    private final String path;
    
    /**
     * Whether to format the output with indentation.
     */
    private final boolean indent;
    
    /**
     * Whether to include newlines in the output.
     */
    private final boolean newline;
    
    /**
     * Whether to include spaces in the output.
     */
    private final boolean space;
    
    /**
     * Get the path to retrieve.
     * @return The path string
     */
    public String getPath() {
        return path;
    }
    
    /**
     * Check if indentation is enabled.
     * @return true if indentation is enabled
     */
    public boolean isIndent() {
        return indent;
    }
    
    /**
     * Check if newlines are enabled.
     * @return true if newlines are enabled
     */
    public boolean isNewline() {
        return newline;
    }
    
    /**
     * Check if spaces are enabled.
     * @return true if spaces are enabled
     */
    public boolean isSpace() {
        return space;
    }
}

File: go/api/server-modules/glidejson/options/json_get_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
package options

const (
	indentKeyword   = "INDENT"   // ValKey API string to designate INDENT
	newlineKeyword  = "NEWLINE"  // ValKey API string to designate NEWLINE
	spaceKeyword    = "SPACE"    // ValKey API string to designate SPACE
	noescapeKeyword = "NOESCAPE" // ValKey API string to designate NOESCAPE
)

// This struct represents the optional arguments for the JSON.GET command.
type JsonGetOptions struct {
	paths    []string
	indent   string
	newline  string
	space    string
	noescape bool
}

func NewJsonGetOptionsBuilder() *JsonGetOptions {
	return &JsonGetOptions{}
}

// Sets the list of paths within the JSON document.
func (jsonGetOptions *JsonGetOptions) SetPaths(paths []string) *JsonGetOptions {
	jsonGetOptions.paths = paths
	return jsonGetOptions
}

// Sets an indentation string for nested levels.
func (jsonGetOptions *JsonGetOptions) SetIndent(indent string) *JsonGetOptions {
	jsonGetOptions.indent = indent
	return jsonGetOptions
}

// Sets a string that's printed at the end of each line.
func (jsonGetOptions *JsonGetOptions) SetNewline(newline string) *JsonGetOptions {
	jsonGetOptions.newline = newline
	return jsonGetOptions
}

// Sets a string that's put between a key and a value.
func (jsonGetOptions *JsonGetOptions) SetSpace(space string) *JsonGetOptions {
	jsonGetOptions.space = space
	return jsonGetOptions
}

// Allowed to be present for legacy compatibility and has no other effect.
func (jsonGetOptions *JsonGetOptions) SetNoescape(noescape bool) *JsonGetOptions {
	jsonGetOptions.noescape = noescape
	return jsonGetOptions
}

// Converts JsonGetOptions into a []string.
func (opts JsonGetOptions) ToArgs() ([]string, error) {
	args := []string{}
	var err error

	if opts.indent != "" {
		args = append(args, indentKeyword, opts.indent)
	}

	if opts.newline != "" {
		args = append(args, newlineKeyword, opts.newline)
	}

	if opts.space != "" {
		args = append(args, spaceKeyword, opts.space)
	}

	if opts.noescape {
		args = append(args, noescapeKeyword)
	}

	if len(opts.paths) > 0 {
		args = append(args, opts.paths...)
	}
	return args, err
}
